services/rsources/stats_collector.go

```diff
@@ -0,0 +1,163 @@
+package rsources
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	uuid "github.com/gofrs/uuid"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/tidwall/gjson"
+)
+
+// StatsCollector collects and publishes stats as jobs are
+// being created, processed and their statuses are being updated.
+type StatsCollector interface {
+
+	// JobsStored captures incoming job statistics
+	JobsStored(jobs []*jobsdb.JobT)
+
+	// JobsStoredWithErrors captures incoming job statistics
+	JobsStoredWithErrors(jobs []*jobsdb.JobT, failedJobs map[uuid.UUID]string)
+
+	// BeginProcessing prepares the necessary indices in order to
+	// be ready for capturing JobStatus statistics
+	BeginProcessing(jobs []*jobsdb.JobT)
+
+	// JobStatusesUpdated captures outgoing job statistics.
+	// A call to BeginProcessing must preceed a call to this method,
+	// so that all necessary indices can been created, since a JobStatus
+	// doesn't carry all necessary job metadata such as jobRunId, taskRunId, etc.
+	JobStatusesUpdated(jobStatuses []*jobsdb.JobStatusT)
+
+	// Publish publishes statistics
+	Publish(ctx context.Context, tx *sql.Tx) error
+}
+
+// NewStatsCollector creates a new stats collector
+func NewStatsCollector(service JobService) StatsCollector {
+	return &statsCollector{
+		service:            service,
+		jobIdsToStatsIndex: map[int64]*Stats{},
+		statsIndex:         map[statKey]*Stats{},
+	}
+}
+
+type statKey struct {
+	jobRunId string
+	JobTargetKey
+}
+
+var _ StatsCollector = (*statsCollector)(nil)
+
+type statsCollector struct {
+	processing         bool
+	service            JobService
```

Comment:
It doesn't seem that you really need a `JobService` here but just a subset of it (i.e. a single method which is `IncrementStats`). Might be a case where you could need a specific interface like `StatsIncrementer` and have it embedded in `JobService`.